subsys/jwt/jwt.c

```diff
@@ -154,23 +159,11 @@ static struct json_obj_descr jwt_payload_desc[] = {
  */
 static void jwt_add_header(struct jwt_builder *builder)
 {
-	static const struct jwt_header head = {
-		.typ = "JWT",
-#ifdef CONFIG_JWT_SIGN_RSA
-		.alg = "RS256",
-#endif
-#ifdef CONFIG_JWT_SIGN_ECDSA
-		.alg = "ES256",
-#endif
-	};
+	int jwt_header_len = ARRAY_SIZE(jwt_header) - 1;
 
-	int res = json_obj_encode(jwt_header_desc, ARRAY_SIZE(jwt_header_desc),
-				  &head, base64_append_bytes, builder);
-	if (res != 0) {
-		/* Log an error here. */
-		return;
-	}
-	base64_flush(builder);
+	strncpy(builder->buf, jwt_header, builder->len - 1);
```

Comment:
This does not safely append to the buffer. The `strncpy` will copy the bytes as long as there is room (possibly leaving off a trailing nul, causing the string to be unterminated), but then the two lines below will change the buffer by the header length, regardless of whether it fits or not.

The old routine would silently fail if there isn't room for the header. Perhaps that could be fixed, but this should be fixed so that at least 1. the resulting string is always null terminated, and 2. if it overflows, the overflow flag should be set in the builder, and 3. the length and buf are managed properly. See `base64_outch` for how this is supposed to be handled.

Reply:
@d3zd3z updated